File: 102/bin_tree_level_traversal.py
	# Definition for a binary tree node.
class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
	def __init__(self, x, left, right):
		self.val = x
		self.left = left
		self.right = right

class Solution(object):
	def levelOrder(self, root):
		"""
		:type root: TreeNode
		:rtype: List[List[int]]
		"""
		h = self.height(root)
		ret = [[] for i in range(h)]

		self.__level_order__(root, 1, h, ret)
		return ret

	def __level_order__(self, root, depth, maxdepth, ret):
		if root == None:
			return

		if root.left == None and root.right == None:
			ret[depth-1].append(root.val)
			return

		if root.left != None and root.right != None:
			self.__level_order__(root.left, depth+1, maxdepth, ret)
			ret[depth-1].append(root.val)
			self.__level_order__(root.right, depth+1, maxdepth, ret)
			return

		if root.left == None:
			ret[depth-1].append(root.val)
			self.__level_order__(root.right, depth+1, maxdepth, ret)
			return

		if root.right == None:
			self.__level_order__(root.left, depth+1, maxdepth, ret)
			ret[depth-1].append(root.val)
			return

	def height(self, root):
		if root == None:
			return 0

		if root.left == None and root.right == None:
			return 1

		if root.left == None:
			return 1 + self.height(root.right)

		if root.right == None:
			return 1 + self.height(root.left)    	

		return 1 + max([self.height(root.left), self.height(root.right)])


if __name__=="__main__":

	# t = TreeNode(4, TreeNode(2, None, None), TreeNode(6, TreeNode(5, None, None), TreeNode(7, None, None)))
	t = TreeNode(4, TreeNode(2, TreeNode(1, None, None), TreeNode(3, None, None)), TreeNode(6, TreeNode(5, None, None), None))
	s = Solution()
	print s.height(t)
	print s.levelOrder(t)

File: 102/BinTreeLevelTraversal.java
import java.lang.Math;
import java.util.List;
import java.util.ArrayList;

public class BinTreeLevelTraversal {

	void get_level(TreeNode t, int i, int lvl, List<List<Integer>> ret) {
		assert i >= 0;
		assert t != null;
		assert ret != null;
		assert ret.get(i) != null;
		
		if (i >= lvl-1) {
			ret.get(i).add(t.val);
			return;
		}
		if (t.left != null) get_level(t.left, i+1, lvl, ret);
		if (t.right != null) get_level(t.right, i+1, lvl, ret);
	}

	int height(TreeNode t) {
		if (t == null) return 0;
		return 1+Math.max(height(t.left), height(t.right));
	}

	void __level_order__(TreeNode t, List<List<Integer>> ret) {
		int h = height(t);
		for (int i=0; i<h; i++) {
			ret.add(new ArrayList<Integer>());
			get_level(t, 0, i+1, ret);
		}
	}

	List<List<Integer>> level_order(TreeNode t) {
		List<List<Integer>> ret = new ArrayList<List<Integer>>();
		__level_order__(t, ret);
		return ret;
	}

	// solution 2: add to ret as traversing inorder
	List<List<Integer>>
	traverse(TreeNode t) {
		List<List<Integer>> ret = new ArrayList<List<Integer>>();
		__traverse__(t, 0, ret);
		return ret;
	}
	void __traverse__(TreeNode t, int lvl, List<List<Integer>> ret) {
		if (t == null) return;
		if (ret.size() <= lvl)
			ret.add(new ArrayList<Integer>());

		if (t.left != null) __traverse__(t.left, lvl+1, ret);
		ret.get(lvl).add(t.val);
		if (t.right != null) __traverse__(t.right, lvl+1, ret);
	}
}